tools/promql-generator/promql-generator.go

```diff
@@ -0,0 +1,354 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+/*
+Promql generator generates random PromQL queries for given label sets.
+
+The label sets can be statically defined in a configuration file or can be obtained by connecting to a running Prometheus server.
+
+The PromQL queries are generated using the promqlsmith library.
+
+The generated queries are written to stdout, one query per line.
+
+Usage:
+
+promql-generator -series.source=file -labels.file=<labels_file.js> [-promql.count=<int>]
+
+or
+
+promql-generator -series.source=prometheus -prometheus.address=<url> -prometheus.matcher=<matcher> [-promql.count=<int>]
+
+The flag variables are;
+
+  - -labels.file=<labels_file.js> (required for -series.source=file, default=sample_series.js) - a file which contains a json representation of 1 or more label sets. An example file is provided.
+  - -prometheus.address=<url> (required for -series.source=prometheus, default=http://127.0.0.1:9009/prometheus) - a fully qualified URL to a Prometheus server.
+  - -prometheus.matcher=<matcher> (required for -series.source=prometheus, default={job="prometheus"}) - a matcher to limit the label sets discovered from the Prometheus server.
+  - -promql.count=<int> (optional default=100) - the number of queries generated. 0 < count < 10000
+*/
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"math/rand"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/cortexproject/promqlsmith"
+	"github.com/efficientgo/core/errors"
+	"github.com/grafana/dskit/flagext"
+	"github.com/prometheus/client_golang/api"
+	v1 "github.com/prometheus/client_golang/api/prometheus/v1"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/promql/parser"
+	"golang.org/x/exp/slices"
+)
+
+var (
+	// add any functions to this map to exclude them from the generated queries
+	unsupportedFunctions = map[string]struct{}{}
+
+	enabledBinops = []parser.ItemType{
+		parser.SUB,
+		parser.ADD,
+		parser.MUL,
+		parser.MOD,
+		parser.DIV,
+		parser.EQLC,
+		parser.NEQ,
+		parser.LTE,
+		parser.GTE,
+		parser.LSS,
+		parser.GTR,
+		parser.POW,
+		parser.LAND,
+		parser.LOR,
+		parser.LUNLESS,
+	}
+)
+
+// A configOption holds a definition of a command line flag option
+type configOption[T string | int] struct {
+	// The CLI flag string
+	cliFlag string
+	// A usage description
+	usage string
+	// A default value
+	defaultValue T
+	// The value user input once parsed
+	flag *T
+}
+
+// A configOptionInt is an extension of configOption and adds in a maxValue attribute
+type configOptionInt struct {
+	config   configOption[int]
+	maxValue int
+}
```

Comment:
This is quite different to how we handle configuration flags elsewhere - is there a particularly strong reason to diverge from the existing pattern rather than defining a type for config flags like this?